Contribution guide (#2840)
prometheus/docs: docs/guides/contributing.md

Slack](https://slack.cncf.io/). The main development channel is
`#prometheus-dev`, but there are plenty of specialized channels. Look for
channel names like `#prometheus-...-dev`, e.g. `#prometheus-protobuf-dev`.

Note that Slack is a silo. The content is not indexed by external search
engines, there is no easy way to export and archive the content, and even
read-only access requires a login. Therefore, consider everything in Slack as
ephemeral and inaccessible for the general public. Important information, like
the result of a discussion, should also be published via other channels (e.g.
on [GitHub](github) or the [mailing list](#developer-mailing-list)) to make it
accessible and durable. Avoid linking to Slack messages from other media
without summarizing the content of the linked messages.

### IRC

The IRC development channel is `#prometheus-dev` on
[irc.libera.chat](https://libera.chat/). It is also accessible via Matrix as
[`#prometheus-dev:matrix.org`](https://app.element.io/#/room/#prometheus-dev:matrix.org).
In principle, IRC/Matrix is the preferred chat-like communication channel
because it is much more open than Slack. However, in practice the traffic on

> IRC is very low.

Author: I removed it for now. No reason to add more information here that has limited value for the task at hand.

---
title: Contributing
---

# Contributor's guide

This document provides general guidelines for contributors of all experience
levels.

If you are new, you might want to check out [how to pick something to
work on](#how-to-pick-something-to-work-on) first.

If you are an experienced contributor Looking to contribute a change, check the 
[high level process](#how-to-get-a-pr-merged).

If you have concrete plans for something more involved already, please refer to the [proposal
process](#proposal-process).

This document is guide for the prometheus/prometheus repository. Much of this 
applies to other repos in the Prometheus community, but these projects might 
have their own guides and rules. For repository-specific information, check out 
the README.md and CONTRIBUTING.md files of each repository.

## How to get a PR merged

If you’re working on your first contribution please read this whole document, this section is aimed at experienced Open Source contributors.

- Test changes locally and consider adding tests for your change.

> - Commit with `git commit -s` to sign the DCO.

Reviewer: I usually run some formatting and linting locally to avoid cycles.

```suggestion
- To avoid lint failures in CI it's useful to run `make common-format` and `make lint` and fix issues.
- Commit with `git commit -s` to sign the DCO.
```

best judgment to make a logical series of changes. Good commits make reviews 
easier and thus can be merged faster. For example it's a good pattern to have a 
commit adding a test to expose a bug and then have a separate commit to fix the 
bug.
Commit messages must include a `Signed-off-by: <author identity>` line. With 
this the author agrees to the terms published at 
https://developercertificate.org/ for _that_ particular contribution.

Once you have a change you would like to propose, push it to your personal fork 
of Prometheus and open a pull request against the default branch. The default 
branch is usually `main` but can be `master` in some repositories.
Some situations require a PR against a release branch, e.g. `release-3.5`. The 
most common situations are fixes for a release candidate for a new release or a 
fix for a LTS version of Prometheus. If in doubt, ask via one of the channels 
mentioned in this document.

The needed reviewers will be added automatically. Anyone else that should or 
wants to review a change can be mentioned by their user name in a comment, but 
please avoid pinging random community members.

> We have checks that run on every PR. To merge all checks should succeed. Any 

Reviewer 2: ```suggestion
We have checks that run on every PR. To merge, all checks should succeed. Any 
```

failing checks should be investigated and addressed by the PR author.

> Reviewers might request changes, which requires the author to either add commits 

Reviewer: Personal preference: I generally don't look at individual commits (unless the author calls it out).
If a PR is too big, I'd rather it's split into multiple PRs. Also if half the PR is renaming things and shuffling things around, and the other half is logic changes, I'm sure to request splitting into an easy to review refactor PR and the substantial PR. 

I think it might discourage people from contributing if we're strict about commits. I'd rather we say something on the PR complexity.

- [2021 virtual developer summit 1 notes](https://docs.google.com/document/d/10o4gkjgK46MdUHTowpUG_pyeON3Z5k0CbPBqQ211KOE)
- [2021-2024 developer summit rolling notes](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE)

### Facilitator

The Facilitator role was created to help the Prometheus team to run the
Developer Summits effectively. It's a rotational role (switches for
every meeting) and its responsibilities are spread across different
phases of the summit:

#### Before the summit

Before the summit, the Facilitator's main goal is to help the
Prometheus team define the agenda and the topics to be discussed while
making sure interested parties of the most voted topics will be able to
attend the summit. We suggest the following tasks:

- Two or three days before the meeting, send reminders in our public
  community channels inviting people to add Agenda Topics, and
  Prometheus Team members and maintainers to vote on topics they'd
  like to discuss.
- One day before the meeting, reach out to "Topic owners" who
  received the most votes to make sure they'll make it to the summit.

#### During the summit

> During the summit, the Facilitator is here to make sure the meeting runs

Reviewer: ```suggestion
During the summit, the Facilitator is there to make sure that the meeting runs
```

the result of a discussion, should also be published via other channels (e.g.
on [GitHub](github) or the [mailing list](#developer-mailing-list)) to make it
accessible and durable. Avoid linking to Slack messages from other media
without summarizing the content of the linked messages.

### Developer mailing list

The
[prometheus-developers](https://groups.google.com/forum/#!forum/prometheus-developers)
mailing list
([mirror](https://www.mail-archive.com/prometheus-developers@googlegroups.com/))
is suitable for announcements and more formal discussions about overarching
topics. The mailing list archive is indexed by search engines and thus a good
way to find past discussions and ensure that information is not lost in a silo
like Slack.

### Developer summits

Developer summits are public meetings to discuss more involved development
topics. For the schedule, meeting notes, and many other details see the
[dedicated section](#developer-summit-details) below.

### Work groups

If developers working on a specific topic would like to conduct online meetings
on a more or less regular basis, they start a work group. Work group meetings
are public and published via the [Prometheus

> calendar](https://calendar.google.com/calendar/u/0/embed?src=prometheus.io_bdf9qgm081nrd0fe32g3olsld0%40group.calendar.google.com).

Reviewer 3: It might be worth mentioning that anyone can join to just listen, and we encourage participation even from folks who are not officially part of the team.
